### ISSUE TITLE
TLS performance enhancement

### ISSUE THREAD

**Issue by alk** [gperftools/gperftools#84]
Originally reported on Google Code with ID 81

```
libtcmalloc.so hits __tls_get_addr heavily. An oprofile of a workload that
uses new a lot shows this:

%        app name                  symbol name
10.7240  libtcmalloc_orig.so.0.0.0 operator new[](unsigned long)
 8.4210  ld-2.5.so                 __tls_get_addr

This happens because we are using __thread in an -fPIC compiled library.
Thanks to a tip from Alan Modra, we can fix this by using the initial exec
tls model. Page 34 of the following document explains it:

http://people.redhat.com/drepper/tls.pdf

Basically this optimisation can only be used in shared libraries that are
never dlopened. It would be a bug to dlopen a malloc replacement (because
we'd end up calling free on objects that were allocated by the previous
malloc) so this should be OK.

With this fix the benchmark in question improved by 18%
```

Reported by `anton.blanchard` on 2008-10-15 09:24:36

<hr>
- _Attachment: [tls_optimisation.patch](https://storage.googleapis.com/google-code-attachments/gperftools/issue-81/comment-0/tls_optimisation.patch)_
